# Assembly a bacterial genome using fastq reads generated by Oxford Nanopor MiniON platform.

 **After prepare the sequencing library and run it on a MiniON flowcell, [Guppy](https://nanoporetech.com/nanopore-sequencing-data-analysis) was used for basecalling. 
 A subset of six fastq files were obtained of this sequencing experiment. The following protocol describes the steps for assembly those reads into a bacterial genomes**
 
 ### Obtaining the fastq files
 
 1. Loggin to Orion cluster

```bash
ssh bio326-21-0@login.orion.nmbu.no
```

2. Go to the $SCRATCH folder. 
```bash
[bio326-21-0@login ~]$ cd $SCRATCH/
```
3. Create a directory named GenomeAssemblyBio326 and copy from (/mnt/SCRATCH/bio326-21/GenomeAssembly/) the tarball (SalmonBacteria.rawReads.subset.tar.gz) with the fastq files to your $SCRATCH/GenomeAssemblyBio326 directory.

```bash
[bio326-21-0@login bio326-21-0]$ mkdir GenomeAssemblyBio326
[bio326-21-0@login bio326-21-0]$ cd GenomeAssemblyBio326/
[bio326-21-0@login GenomeAssemblyBio326]$ cp /mnt/SCRATCH/bio326-21/GenomeAssembly/SalmonBacteria.rawReads.subset.tar.gz .
[bio326-21-0@login GenomeAssemblyBio326]$ ls
SalmonBacteria.rawReads.subset.tar.gz
```
*If you see the SalmonBacteria.rawReads.subset.tar.gz in your new GenomeAssembly directory means you have succeded copy this tarball.*

4. Decompress the tarball to access the files. For this, we can use the command tar and the options -x (decompress) z (gzip file) v (verbose or print the results) f (file).

```bash
[bio326-21-0@login GenomeAssemblyBio326]$ tar -xzvf SalmonBacteria.rawReads.subset.tar.gz
SalmonBacteria.rawReads.subset/
SalmonBacteria.rawReads.subset/fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
SalmonBacteria.rawReads.subset/fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
SalmonBacteria.rawReads.subset/fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
SalmonBacteria.rawReads.subset/fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
SalmonBacteria.rawReads.subset/fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
SalmonBacteria.rawReads.subset/fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
```

5. Enter to the SalmonBacteria.rawReads.subset directory and take a look:

```bash
[bio326-21-0@login GenomeAssemblyBio326]$ cd SalmonBacteria.rawReads.subset
[bio326-21-0@login SalmonBacteria.rawReads.subset]$ ls
fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq  fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
```
**As you notice there are six fastq files in this directory. It is often useful to concatenate all the different fastq files into one big file for downstream analyses.**

```bash
[bio326-21-0@login SalmonBacteria.rawReads.subset]$ cat *.fastq > SalmonBacteria.total.fastq
```

Now you will have a "big" (> 100 Mb) fastq file (SalmonBacteria.total.fastq) with all the sequenced reads:

```bash
[bio326-21-0@login SalmonBacteria.rawReads.subset]$ ls -lrth
total 398M
-rwxrwxr-x 1 bio326-21-0 bio326-21-0  32M Mar 22 16:17 fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_1_0.fastq
-rwxrwxr-x 1 bio326-21-0 bio326-21-0  32M Mar 22 16:17 fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_0_0.fastq
-rwxrwxr-x 1 bio326-21-0 bio326-21-0  37M Mar 22 16:17 fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_5_0.fastq
-rwxrwxr-x 1 bio326-21-0 bio326-21-0  35M Mar 22 16:17 fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_4_0.fastq
-rwxrwxr-x 1 bio326-21-0 bio326-21-0  34M Mar 22 16:17 fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_3_0.fastq
-rwxrwxr-x 1 bio326-21-0 bio326-21-0  32M Mar 22 16:17 fastq_runid_cbaffd65431ed3590f2402612142061571365f8a_2_0.fastq
-rw-rw-r-- 1 bio326-21-0 bio326-21-0 199M Mar 22 17:08 SalmonBacteria.total.fastq
```

We can count the number of reads in this "big" concatenated file. For this we can count the number of lines of the file and divide by four. Why by four? 

```
A FASTQ file normally uses four lines per sequence.

Line 1 begins with a '@' character and is followed by a sequence identifier and an optional description (like a FASTA title line).
Line 2 is the raw sequence letters.
Line 3 begins with a '+' character and is optionally followed by the same sequence identifier (and any description) again.
Line 4 encodes the quality values for the sequence in Line 2, and must contain the same number of symbols as letters in the sequence.
```




 
